server/test/unit/com/thoughtworks/go/server/persistence/PipelineRepositoryTest.java

```diff
@@ -89,15 +116,89 @@ public void shouldCachePipelineSelectionForGivenId() throws Exception {
 
     @Test
     public void shouldInvalidatePipelineSelectionCacheOnSaveOrUpdate() throws Exception {
-
-
         PipelineSelections pipelineSelections = new PipelineSelections();
         pipelineSelections.setId(1);
         pipelineSelections.update(new ArrayList<String>(), new Date(), 2L, true);
         pipelineRepository.saveSelectedPipelines(pipelineSelections);
 
-
         verify(goCache).remove(pipelineRepository.pipelineSelectionForCookieKey(1L));
         verify(goCache).remove(pipelineRepository.pipelineSelectionForUserIdKey(2L));
     }
+
+    @Test
+    public void shouldUpdateTimelineEntriesAndEntriesForRollback() throws Exception {
+        Object[] pipelineRow1 = {"p1", new BigInteger("1"), new BigInteger("1"), new Date(), "fingerprint", 1.0, "r1", null, new BigInteger("1"), new BigInteger("1")};
+        Object[] pipelineRow2 = {"p1", new BigInteger("2"), new BigInteger("2"), new Date(), "fingerprint", 2.0, "r2", null, new BigInteger("1"), new BigInteger("1")};
+
+        stubPipelineInstancesInDb(pipelineRow1, pipelineRow2);
+        ArrayList<PipelineTimelineEntry> tempEntries = new ArrayList<>();
+        PipelineTimeline pipelineTimeline = new PipelineTimeline(pipelineRepository, transactionTemplate, transactionSynchronizationManager);
+
+        pipelineRepository.updatePipelineTimeline(pipelineTimeline, tempEntries);
+
+        PipelineTimelineEntry timelineEntry1 = pipelineTimeline.getEntryFor(new CaseInsensitiveString("p1"), 1);
+        PipelineTimelineEntry timelineEntry2 = pipelineTimeline.getEntryFor(new CaseInsensitiveString("p1"), 2);
+        assertThat(pipelineTimeline.instanceCount(new CaseInsensitiveString("p1")), is(2));
+        assertNotNull(timelineEntry1);
+        assertNotNull(timelineEntry2);
+        assertThat(tempEntries.size(), is(2));
+        assertThat(tempEntries, containsInAnyOrder(timelineEntry1, timelineEntry2));
+    }
+
+    @Test
+    public void shouldNotUpdateTimelineEntriesAndEntriesForRollbackUponFailureDuringRetrieval() throws Exception {
+        Object[] pipelineRow1 = {"p1", new BigInteger("1"), new BigInteger("1"), new Date(), "fingerprint", 1.0, "r1", null, new BigInteger("1"), new BigInteger("1")};
+        Object[] pipelineRow2 = {"p1", "2abcd", new BigInteger("2"), new Date(), "fingerprint", 2.0, "r2", null, new BigInteger("1"), new BigInteger("1")};
```

Comment:
Can you change "2abcd" to something which describes what's happening? Something like "WILL-CAUSE-FAILURE-DURING-RETRIEVAL"?